db/migrate/20231210061330_add_registered_at_to_fts_targets_with_index.rb

```diff
@@ -0,0 +1,104 @@
+  def up
+    return if !table_exists?(:fts_targets)
+
+    add_column :fts_targets, :registered_at, :timestamp
+
+    ActiveRecord::Base.transaction do
+      update_registered_at_using_last_modified_at
+      update_registered_at_using_created_on
+    end
+
+    if Redmine::Database.mysql?
+      add_index :fts_targets, :registered_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :registered_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+  end
+
+  def down
+    return if !table_exists?(:fts_targets)
+
+    if Redmine::Database.mysql?
+      remove_index :fts_targets, :registered_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+
+    remove_column :fts_targets, :registered_at
+  end
+
+  private
+
+  class FtsTarget < ActiveRecord::Base
+  end
+
+  class FtsType < ActiveRecord::Base
+  end
+
+  class Issue < ActiveRecord::Base
+  end
+
+  class Message < ActiveRecord::Base
+  end
+
+  class Project < ActiveRecord::Base
+  end
+
+  class WikiPage < ActiveRecord::Base
+  end
+
+  def update_registered_at_using_last_modified_at
+    type_names = ['Attachment', 'Change', 'Changeset', 'CustomValue', 'Document', 'Journal', 'News']
+
+    FtsTarget.where(source_type_id: FtsType.where(name: type_names).select(:id))
+             .update_all(registered_at: FtsTarget.arel_table[:last_modified_at])
+  end
+
+  def update_registered_at_using_created_on
+    types = [Issue, Message, Project, WikiPage]
+    types.each do |type|
+      subquery_for_created_on = type.where(type.arel_table[:id].eq(FtsTarget.arel_table[:source_id]))
+                                    .select(:created_on)
+                                    .to_sql
+      FtsTarget.where(source_type_id: FtsType.where(name: type.table_name.singularize.camelize).select(:id))
```

Comment:
Can we use `type.name.split("::").last`?
(Or does Active Support provide a convenient method?)

Reply:
fix: [4248844](https://github.com/clear-code/redmine_full_text_search/pull/118/commits/4248844d40916eafab1fc0c75f557bdef8ef1ca5): There is the `demodulize` method although I don't know that.
ref: https://api.rubyonrails.org/classes/String.html#method-i-demodulize